README.md
# jQuery JumboSlider Plugin

Yet another responsive slider for any kind of content.

[![Built with Grunt](http://cdn.gruntjs.com/builtwith.svg)](https://gruntjs.com/) [![GitHub license](https://img.shields.io/badge/license-MIT-yellow.svg)](https://raw.githubusercontent.com/martinmethod/jumboslider/prod/LICENSE-MIT) [![Travis](https://img.shields.io/travis/martinmethod/jumboslider.svg)](https://travis-ci.org/martinmethod/jumboslider) [![David](https://img.shields.io/david/dev/martinmethod/jumboslider.svg)](https://david-dm.org/martinmethod/jumboslider?type=dev) [![GitHub release](https://img.shields.io/github/release/martinmethod/jumboslider.svg)](https://github.com/martinmethod/jumboslider/releases/latest) [![npm](https://img.shields.io/npm/v/jumboslider.svg)](https://www.npmjs.com/package/jumboslider) [![Bower](https://img.shields.io/bower/v/jumboslider.svg)](https://github.com/martinmethod/jumboslider)

## Description

JumboSlider is a light and responsive jQuery plugin that provides a simple way to transform any kind of content on any web page into a compact and beautiful slideshow. It's easy to setup with the given options and it's also a great module to interact with due to its methods and callback events.

## Demo

<a href="https://jumboslider.metodiev.dev" target="_blank">jumboslider.metodiev.dev</a>

## Getting Started

You can [download the plugin as an archive][zip].

[zip]: https://github.com/martinmethod/jumboslider/zipball/prod

Or you can grab it by using **npm**:

```javascript
npm install jumboslider
```

Or you can grab it by using **Bower**:

```javascript
bower install jumboslider
```

## Installation

Include the script after the jQuery library (unless you package scripts otherwise):

```html
<script src="/path/to/jumboslider.min.js"></script>
```

Also include the stylesheet for the plugin:

```html
<link type="text/css" rel="stylesheet" href="/path/to/jumboslider.min.css">
```

## Usage

### Initialization

Firstable, you need to have a proper markup:

```html
<div class="jumboslider">
    <div class="jumboslider-viewport">
        <div class="jumboslider-overview">
            <div class="jumboslider-item">
                <!-- content of your choice -->
            </div>
            <div class="jumboslider-item">
                <!-- content of your choice -->
            </div>
            <div class="jumboslider-item">
                <!-- content of your choice -->
            </div>
            <div class="jumboslider-item">
                <!-- content of your choice -->
            </div>
        </div>
    </div>
</div>
```

Now, there are two ways of initializing JumboSlider.

#### I. jQuery function

The first way is as follows:

```javascript
$.jumboslider(); // returns the target
```

This is the most basic initialization. By calling it so, the plugin will look for any element that has a "**jumboslider**" class. If no such elements, nothing will happen.

Of course, you can also provide specific target/s with a custom selector by doing so:

```javascript
$.jumboslider({
    target: $('selector')
});
```

#### II. Method function

The second way is as an object method. So, we can also use the plugin this way:

```javascript
$('selector').jumboslider(); // returns the target
```

### Options

#### target

Defines a custom target selector.

*The target property is only available when initialize JumboSlider as a jQuery function like the example below.*

```javascript
$.jumboslider({
    target: $('selector')
});
```

Default value: **$('.jumboslider')**

#### startPosition

Defines which item will be the first one to be shown.

```javascript
$.jumboslider({
    startPosition: 1
});
```

Type: **number**  
Default value: **1**  

#### arrows

Defines whether or not graphic arrows should be shown

```javascript
$.jumboslider({
    arrows: true
});
```

Type: **boolean**  
Default value: **true**  

#### pagination

Defines whether or not a pagination should be shown

```javascript
$.jumboslider({
    pagination: true
});
```

Type: **boolean**  
Default value: **true**  

#### transition

Defines the speed of the transition between slides (in milliseconds).

```javascript
$.jumboslider({
    transition: 200
});
```

Type: **number**  
Default value: **500**  

#### loop

Defines whether or not the JumboSlider should be able to start over from the first item after it reaches to the last one and also to jump back to the last item right from the first one.

```javascript
$.jumboslider({
    loop: false
});
```

Type: **boolean**  
Default value: **false**  

#### autoplay

Defines whether or not the JumboSlider should runs automatically by providing a lifetime duration for the slides (in milliseconds). The default duration is `0` which means the autoplay is off.

```javascript
$.jumboslider({
    autoplay: 3000
});
```

Type: **number**  
Default value: **0**  

### Events

There are two ways of binding JumboSlider events.

**1.** As option properties:

```javascript
$('selector').jumboslider({
    onSlide: function(e, target) {
        // do something
    }
});
```

**2.** As jQuery events:

```javascript
$('selector').jumboslider().bind('onSlide', function(e, target) {
    // do something
});
```

And here are all available events:

#### onSlide

Triggers when a slide transition is being fired.

```javascript
var slider = $('selector').jumboslider();

slider.bind('onSlide', function(e, target) {
    // do something
});
```

#### onArrowClick

Triggers when a graphic arrow controller is being used.

```javascript
var slider = $('selector').jumboslider();

slider.bind('onArrowClick', function(e, target) {
    // do something
});
```

#### onPaginationClick

Triggers when a pagination controller is being used.

```javascript
var slider = $('selector').jumboslider();

slider.bind('onPaginationClick', function(e, target) {
    // do something
});
```

### Methods

#### slideNext()

Slides to the next item (if any).

```javascript
var slider = $('selector').jumboslider();

slider.slideNext();
```

#### slidePrev()

Slides to the previous item (if any).

```javascript
var slider = $('selector').jumboslider();

slider.slidePrev();
```

#### slideTo(position)

Slides to a specific item.

```javascript
var slider = $('selector').jumboslider();

slider.slideTo(5); // will slide to the fifth item
```

## Browsers compatibility

- Apple Safari
- Google Chrome
- Microsoft Internet Explorer 9+
- Mozilla Firefox
- Opera

## Dependencies

- [jQuery][jq]

[jq]: https://github.com/jquery/jquery.git

## License

Copyright © 2017 Martin Metodiev. Licensed under the MIT license. [See here for more details.][licence]

[licence]: https://raw.github.com/martinmethod/jumboslider/prod/LICENSE-MIT


CHANGELOG.md
<a name="1.5.0"></a>
# 1.5.0 (2020-01-25)



<a name="1.4.5"></a>
## 1.4.5 (2017-06-02)


### Bug Fixes

* **Silent focus:** Focus slider on mouse enter ([282167b](https://github.com/martinmethod/jumboslider/commit/282167b))



<a name="1.4.0"></a>
# 1.4.0 (2017-05-31)


### Bug Fixes

* **Overview width:** Wrong calculations ([aacd891](https://github.com/martinmethod/jumboslider/commit/aacd891))


### Features

* **Autoplay:** Ability to slide automatically ([b44bc46](https://github.com/martinmethod/jumboslider/commit/b44bc46))
* **Transition:** Ability to define the transition speed ([0d121f0](https://github.com/martinmethod/jumboslider/commit/0d121f0))



<a name="1.3.2"></a>
## 1.3.2 (2017-04-11)


### Bug Fixes

* **githubUsername:** Broken object ([50d06fa](https://github.com/martinmethod/jumboslider/commit/50d06fa))


### Features

* **Banner:** External text configuration ([65588d4](https://github.com/martinmethod/jumboslider/commit/65588d4))



<a name="1.3.1"></a>
## 1.3.1 (2017-04-11)


### Features

* **Arrows:** Arrows are now white ([f814a5b](https://github.com/martinmethod/jumboslider/commit/f814a5b))
* **Build:** External markup support ([ed2b370](https://github.com/martinmethod/jumboslider/commit/ed2b370))



<a name="1.3.0"></a>
# 1.3.0 (2017-04-05)


### Features

* **Deploy:** npm, GitHub Pages & GitHub Releases ([3910056](https://github.com/martinmethod/jumboslider/commit/3910056))



<a name="1.2.0"></a>
# 1.2.0 (2017-04-03)


### Features

* **Building system:** Refactored building system ([3f9ca30](https://github.com/martinmethod/jumboslider/commit/3f9ca30))





src/js/jquery.jumboslider.js
/*!
  jQuery JumboSlider Plugin v1.1.2
  http://jumboslider.martinmetodiev.com

  Copyright (c) 2015 Martin Metodiev
  Licensed under the MIT license.
*/


;(function($) {

'use strict';

// Define all plugin components
var plugin = {
        // Base plugin data
        base: {
            // Default target selector if no such provided
            target: $('.jumboslider'),

            // List of all supported options with their default values
            options: {
                startPosition: 1,
                arrows: true,
                pagination: true,
                loop: false,
                keyboard: true,
                keyboardFocus: false // requires "keyboard" property to be true
                // auto: false,
                // autoDuration: 3000
            }
        },

        events: [
            'onSlide',          // e, data
            'onArrowClick',     // e, data
            'onPaginationClick' // e, data
        ],

        dom: {
            arrows: $('\
                <div class="jumboslider-arrows">\
                    <a href="javascript:;" class="jumboslider-prev-arrow"></a>\
                    <a href="javascript:;" class="jumboslider-next-arrow"></a>\
                </div>\
            '),
            pagination: {
                block: $('<div class="jumboslider-pagination"><div class="holder"></div>'),
                dot: $('\
                    <a href="javascript:;">\
                        <svg version="1.1" class="jumboslider-pagination-mask" \
                        xmlns="http://www.w3.org/2000/svg" \
                        xmlns:xlink="http://www.w3.org/1999/xlink" \
                        x="0px" y="0px" viewBox="0 0 88 64" \
                        style="enable-background:new 0 0 88 64;" xml:space="preserve">\
                            <path fill="white" d="M0,0v64h88V0H0z M44,58c-14.3, \
                            0-26-11.6-26-26S29.7,6,44,6s26,11.6,26,26S58.3,58,44,58z"/>\
                        </svg>\
                    </a>\
                '),
                current: $('<span class="jumboslider-pagination-current">')
            }
        },

        setup: {
            target: function(params) {
                var target = params && params.hasOwnProperty('target') ?
                       params.target : plugin.base.target;

                if (!target.is('.jumboslider')) { target.addClass('jumboslider'); }

                return target;
            },

            options: function(params) {
                var options = {},
                    events = {};

                (function() {
                    for (var option in plugin.base.options) {
                        if (plugin.base.options.hasOwnProperty(option)) {
                            options[option] = params && params.hasOwnProperty(option) ?
                                params[option] : plugin.base.options[option];
                        }
                    }
                }());

                (function() {
                    for (var event = 0; event < plugin.events.length; event++) {
                        if (params && params.hasOwnProperty(plugin.events[event])) {
                            events[plugin.events[event]] = params[plugin.events[event]];
                        }
                    }
                }());

                $.extend(options, {events: events});

                return options;
            }
        },

        methods: {
            init: function(params) {
                var obj = this;
                
                // Attach options
                obj.options = plugin.setup.options(params);

                // Define objects
                obj.viewport = obj.find('.jumboslider-viewport');
                obj.overview = obj.find('.jumboslider-overview');
                obj.items = obj.find('.jumboslider-item');

                // Set keyboard arrows
                if (typeof obj.attr('tabindex') === 'undefined') {
                    obj.attr('tabindex', '1');
                }

                // Setup controllers
                obj.setControllers();

                // Set initial width and position
                obj.setWidth()
                   .setPosition(obj.options.startPosition, 'force');

                // Bind provided events
                for (var event in obj.options.events) {
                    if (obj.options.events.hasOwnProperty(event)) {
                        obj.bind(event, obj.options.events[event]);
                    }
                }

                // Bind window resize event to update jumboslides position & fluid width
                $(window).bind('resize', function() {
                    obj.setWidth()
                       .setPosition(obj.currentPosition, 'force');
                });

                return obj.addClass('jumboslider-ready');
            },

            setControllers: function() {
                var obj = this;
                
                // Keyboard arrows (if active)
                if (obj.options.keyboard) {
                    // Set forced focus (if required)
                    if (obj.options.keyboardFocus) { obj.setKeyboard(); }

                    // Focus the object when the cursor enters or clicks its area
                    obj.bind('mouseenter click', function() { obj.setKeyboard(); });
                }

                // Arrows (if active)
                if (obj.options.arrows && obj.items.length > 1) {
                    obj.createArrows().init();
                }

                // Pagination (if active)
                if (obj.options.pagination && obj.items.length > 1) {
                    obj.createPagination().init();
                }

                return obj;
            },

            calculatePosition: function(direction) {
                var obj = this, pos, elseValue;

                function getElse(direction) {
                    return obj.options.loop ? direction :
                                    obj.currentPosition;
                }

                switch(direction) {
                    case 'prev':
                        elseValue = getElse(obj.items.length);
                        pos = obj.currentPosition - 1 >= 1 ?
                        obj.currentPosition - 1 : elseValue;
                        break;

                    case 'next':
                        elseValue = getElse(1);
                        pos = obj.currentPosition + 1 <= obj.items.length ?
                        obj.currentPosition + 1 : elseValue;
                        break;
                }

                return pos;
            },

            setKeyboard: function() {
                var obj = this;

                function action(e) {
                    var direction;

                    switch(e.keyCode) {
                        case 37: direction = 'prev'; break;
                        case 39: direction = 'next'; break;
                    }

                    obj.setPosition( obj.calculatePosition(direction) );
                }

                if (!obj.is('.jumboslider-focused')) {
                    obj.focus().addClass('jumboslider-focused');
                    obj.bind('keyup', action);

                    obj.bind('blur', function() {
                        obj.unbind('keyup', action)
                           .removeClass('jumboslider-focused');
                    });
                }

                return obj;
            },

            setWidth: function() {
                this.overview.width(((this.items.length + 1) * this.width()) - 1);
                this.items.width(this.width());

                return this;
            },

            setPosition: function(pos, force) {
                var obj = this,
                    validPos = !isNaN(pos) && pos > 0 && pos <= obj.items.length,
                    left = 0,

                    position = {
                        update: function() {
                            obj.previousPosition = obj.currentPosition || null;
                            obj.currentPosition = validPos ? pos : obj.previousPosition;

                            left = '-'+$(obj.items[obj.currentPosition-1])
                                    .position().left+'px';

                            return this;
                        },

                        checkForce: function() {
                            if (force) { obj.addClass('force'); }
                            else { obj.removeClass('force'); }

                            return this;
                        },

                        transit: obj.transit,
                        
                        controllers: function() {
                            if (obj.options.pagination && obj.items.length > 1) {
                                obj.pagination.update();
                            }
                            if (obj.options.arrows && obj.items.length > 1) {
                                obj.arrows.update();
                            }

                            return this;
                        }
                    };

                position.update()
                        .checkForce()
                        .transit(obj.overview, left)
                        .controllers();

                if (obj.previousPosition !== obj.currentPosition) {
                    obj.trigger('onSlide', [obj]);
                }

                return obj;
            },

            createArrows: function() {
                var obj = this,
                    arrows = obj.arrows = plugin.dom.arrows.clone();

                $.extend(arrows, {
                    init: function() {
                        var arrows = this;
                        arrows.arrow = arrows.find('a');
                        arrows.arrowPrev = arrows.find('.jumboslider-prev-arrow');
                        arrows.arrowNext = arrows.find('.jumboslider-next-arrow');

                        arrows.appendTo(obj);
                        arrows.update();
                        arrows.css({opacity: 1});

                        arrows.arrow.bind('click', function(/*e*/) {
                            var newPos;

                            if ($(this).is('.jumboslider-next-arrow')) {
                                newPos = obj.calculatePosition('next');
                            }
                            else if ($(this).is('.jumboslider-prev-arrow')) {
                                newPos = obj.calculatePosition('prev');
                            }

                            obj.trigger('onArrowClick', [obj]);

                            obj.setPosition(newPos);
                        });
                        
                        obj.addClass('jumboslider-arrowed');

                        return arrows;
                    },

                    update: function() {
                        var pos = obj.currentPosition;
                        
                        obj.arrows.arrow.removeClass('hidden-arrow');

                        if (!obj.options.loop) {
                            if (pos === 1) {
                                this.arrowPrev.addClass('hidden-arrow');
                            }
                            else if (pos === obj.items.length) {
                                this.arrowNext.addClass('hidden-arrow');
                            }
                        }

                        return this;
                    }
                });

                return arrows;
            },

            createPagination: function() {
                var obj = this,
                    pagination = obj.pagination = plugin.dom.pagination.block.clone();

                $.extend(pagination, {
                    init: function() {
                        var pagination = this,
                            current = plugin.dom.pagination.current.clone();

                        for (var i = 0; i < obj.items.length; i++) {
                            var dot = plugin.dom.pagination.dot.clone();
                            dot.attr('rel', i+1);
                            pagination.find('.holder').append(dot);
                        }
                        pagination.find('.holder').append(current);

                        pagination.dots = pagination.find('a');
                        pagination.current = current;

                        pagination.appendTo(obj);
                        pagination.update();
                        pagination.css({opacity: 1});

                        pagination.dots.bind('click', function(/*e*/) {
                            if (!$(this).is('.current')) {
                                var newPos = parseInt($(this).attr('rel'));

                                obj.trigger('onPaginationClick', [obj]);

                                obj.setPosition(newPos);
                            }
                        });

                        obj.addClass('jumboslider-paginated');

                        return pagination;
                    },

                    update: function() {
                        obj.pagination.dots.removeClass('current');
                        obj.currentPosition = obj.currentPosition || 1;
                        var currentDot = $(obj.pagination.dots[obj.currentPosition - 1]);

                        obj.transit(pagination.current, currentDot.position().left - 1);
                        currentDot.addClass('current');
                        return this;
                    }
                });

                return pagination;
            },

            transit: function(obj, pos) {
                if (navigator.appVersion.indexOf('MSIE 9.') !== -1) {
                    obj.animate({ left: pos }, 500, function() {
                        // Animation complete.
                    });
                    
                }
                else {
                    obj.css({left: pos});
                }

                return this;
            }
        },

        output: {
            extended: true,
            
            slideNext: function() {
                this.each(function() {
                    var obj = this.jumboslider;

                    obj.setPosition( obj.calculatePosition('next') );
                });

                return this;
            },

            slidePrev: function() {
                this.each(function() {
                    var obj = this.jumboslider;

                    obj.setPosition( obj.calculatePosition('prev') );
                });

                return this;
            },

            slideTo: function(pos, force) {
                this.each(function() {
                    var obj = this.jumboslider;

                    if (pos !== obj.currentPosition && pos <= obj.items.length) {
                        obj.setPosition(pos, force);
                    }
                });

                return this;
            },

            destroy: function() {
                this.each(function() {
                    var obj = this.jumboslider;

                    (function() {
                        if (obj.options.arrows) {
                            obj.removeClass('jumboslider-arrowed');
                            obj.arrows.remove();
                        }

                        if (obj.options.pagination) {
                            obj.removeClass('jumboslider-paginated');
                            obj.pagination.remove();
                        }

                        obj.overview.removeAttr('style');
                        obj.items.removeAttr('style');

                        obj.removeClass('jumboslider-ready force');

                        obj.unbind('keyup');
                        obj.unbind('blur');
                    })();

                    // Unbind provided events
                    for (var event in obj.options) {
                        obj.unbind(event, obj.options[event]);
                    }

                    // Delete object from the targets' list
                    for (var i = 0; i < $.jumboslider.targets.length; i++) {
                        if (this === $.jumboslider.targets[i][0]) {
                            $.jumboslider.targets.splice(i, 1);
                            break;
                        }
                    }

                    // Delete jumboslider object
                    delete this.jumboslider;
                });

                // Delete output properties
                for (var property in plugin.output) {
                    if (this.hasOwnProperty[plugin.output[property]]) {
                        delete this[plugin.output[property]];
                    }
                }

                return this;
            }
        }
    };

// Define plugin as a jQuery function
$.jumboslider = function(params) {
    // Setup target
    var target = plugin.setup.target(params);

    // Create a jQuery Object Instance (extended with jumboslider) inside the DOM object
    target.each(function() {
        if (!this.jumboslider) {
            $.extend(this, {
                jumboslider: $.extend($(this), plugin.methods)
            });

            var jumboslider = this.jumboslider;

            $.jumboslider.targets.push(jumboslider.init(params));
        }
    });

    // Extend target with public methods (if not yet)
    if (!target.extended) { $.extend(target, plugin.output); }

    return target;
};

// Define plugin as a method function
$.fn.jumboslider = function(params) {
    params = $.extend({}, params, {target: $(this)});

    return $.jumboslider(params);
};

// Create public storage array with all DOM elements that are active targets of the plugin
$.jumboslider.targets = [];

}(jQuery));